crates/ironrdp-server/src/builder.rs

```diff
@@ -0,0 +1,128 @@
+use std::{marker::PhantomData, net::SocketAddr};
+use tokio_rustls::TlsAcceptor;
+
+use super::display::RdpServerDisplay;
+use super::handler::RdpServerInputHandler;
+use super::server::*;
+
+pub struct WantsAddr {}
+pub struct WantsSecurity {
+    addr: SocketAddr,
+}
+pub struct WantsHandler {
+    addr: SocketAddr,
+    security: RdpServerSecurity,
+}
+pub struct WantsDisplay<H> {
+    addr: SocketAddr,
+    security: RdpServerSecurity,
+    handler: H,
+}
+pub struct BuilderDone<H, D> {
+    addr: SocketAddr,
+    security: RdpServerSecurity,
+    handler: H,
+    display: D,
+}
+
+pub struct RdpServerBuilder<State, H, D> {
+    state: State,
+    _handler: PhantomData<H>,
+    _display: PhantomData<D>,
+}
```

Comment:
Nice, I really like this kind of type safe builders (like the one in rustls)

Reply:
I agree that compile-time checks are always preferable to runtime checks. I may suggest to use [`typed_builder`](https://docs.rs/typed-builder/latest/typed_builder/derive.TypedBuilder.html) crate for such purposes, It has great error reporting, minimal dependencies, and is easy to use (in my opinion). This will greatly decrease boilerplate code for the builder types.

 I used it on my previous job, this crate proved to be very useful! @CBenoit what do you think about it? Making builder type is as simple as

Reply:
Personally, I still prefer my builder because it works like a state machine. You can easily define multiple setters that take you from one state to the next, and it forces you to always do the steps in the same order.
Like @CBenoit noticed, it was inspired by the [one in rustls](https://docs.rs/rustls/latest/rustls/struct.ConfigBuilder.html).

But I looked into `typed_builder` (thanks for the suggestion @pacmancoder) and it seems like it could replace this entire file with a couple annotations. So if you guys prefer that, I am ok with it.

It would be something like this
```rust
    let mut server = RdpServer::builder()
        .with_options(RdpServerOptions {
            addr: ([127, 0, 0, 1], 3389).into(),
            security: RdpServerSecurity::Tls(tls_acceptor),
        })
        .with_display(Box::new(display_handler))
        .with_handler(Box::new(input_handler))
        .build();
```

Compared to the current one
```rust
     let mut server = RdpServer::builder()
         .with_addr(([127, 0, 0, 1], 3389))
         .with_tls(tls_acceptor)
         .with_input_handler(input_handler)
         .with_display_handler(display_handler)
         .build();
```

Reply:
I just realized that, after changing `RdpServer` to use dynamic dispatch, I can also use it to simplify the builder.
But my point still stands, if you would rather use `typed_builder` I'm ok with it.

Reply:
After further discussion with @pacmancoder on slack, we agreed on merging the current builder and see if `typed-builder` would be used somewhere else before committing to using the crate.
Main reasons are:
- `typed-builder` depends on the `syn` crate, and ideally we want to push this dependency as far as possible from the foundational crates so it doesn’t become too much of a compilation bottleneck (also, this crate often ends up duplicated as it can be seen in the `Cargo.lock` on current master). We’re not yet in this ideal world, but that’s the goal. Note however that I’m open to using some proc-macro crates in `ironrdp-server` since this crate will probably end up either [in the community or extra tier](https://github.com/Devolutions/IronRDP/blob/master/ARCHITECTURE.md#extra-tier), which is "far enough" from the foundational crates per IronRDP standard.
- The handcrafted API is indeed looking a bit better.

So there is no need to change anything immediately.